hal/armv8m/mcx/_init.S

```diff
@@ -0,0 +1,359 @@
+	orr r7, r7, #1
+
+	/* Store hardware saved registers on kernel stack */
+	add r10, r10, #(5 * 4)
+	stm r10, {r6-r8} /* lr, pc, psr */
+
+	/* Copy arguments back to the user stack */
+	stmdb r0!, {r1-r4}
+	mov r1, r0
+	ldrb r0, [r7, #-3]
+
+	/* Prepare pseudo context */
+	mov r7, #0x01000000
+	ldr r6, =syscalls_dispatch
+	ldr r5, =_syscallend
+	stmdb sp!, {r0-r7} /* PSR, PC, LR, R12, R3, R2, R1, R0 */
+
+	/* Exit handler mode to kernel thread mode */
+	ldr lr,=RET_THREAD_MSP
+	bx lr
+
+_syscallend:
+	/* return value in r0 */
+	ldr lr, [sp, #8] /* psp */
+	add lr, lr, #(8 * 4)
+	msr psp, lr
+	isb
+
+	add r10, sp, #((14 * 4) + (5 * 4))
+	ldm r10, {r2,r3} /* lr, pc */
+
+	add lr, sp, #12
+	ldmia lr!, {r4-r11}
+	add lr, lr, #(11 * 4)
+	mov sp, lr
+
+	/* Switch stacks */
+	mov r1, #3
+	msr control, r1
+	isb
+
+	mov lr, r2
+	mov pc, r3
+.size _syscall_dispatch, .-_syscall_dispatch
+.ltorg
+
+.globl _exceptions_dispatch
+.type _exceptions_dispatch, %function
+
+_exceptions_dispatch:
+	cpsid if
+	isb
+
+	mrs r0, psp
+	stmdb sp!, {r0, r4-r11, lr}
+
+	mrs r0, ipsr
+	mov r1, sp
+
+	b exceptions_dispatch
+.size _exceptions_dispatch, .-_exceptions_dispatch
+.ltorg
+
+.globl _interrupts_dispatch
+.type _interrupts_dispatch, %function
+_interrupts_dispatch:
+	mov r0, #0
+	msr control, r0
+	isb
+
+	mov r0, sp
+	and r12, lr, #(1 << 2)
+	cmp r12, #0
+	it ne
+	subne sp, sp, #(8 * 4) /* space for hw-saved ctx */
+	str r0, [sp, #-8]!
+
+	mrs r0, ipsr
+	mrs r3, psp
+	sub r1, sp, #48
+	stmdb sp!, {r1-r11, lr}
+
+	/* if we came from userspace, copy hw-saved regs to kstack
+	 * in case of signal handling
+	 */
+
+	beq _intd0 /* cmp r12, #0 */
+
+	/* psp in r3 - load hw-saved regs */
+	ldm r3, {r4-r11}
+	add r12, sp, #(14 * 4)
+	stm r12, {r4-r11}
+
+_intd0:
+	bl interrupts_dispatch
+	ldr r1, [sp]
+	mov sp, r1
+	isb
+_intd1:
+	ldr lr, [sp, #(11 * 4)]
+
+	and r12, lr, #(1 << 2)
+	cmp r12, #0
+	beq _intd2
+
+	/* userspace return - restore registers in case of signal handling */
+	ldr r3, [sp, #(2 * 4)] /* psp */
+	add r1, sp, #(14 * 4)
+	ldm r1, {r4-r11}
+	stm r3, {r4-r11}
+
+_intd2:
+	ldmia sp!, {r1-r11}
+	ldr r0, [sp, #4]
+	mov sp, r0
+
+	msr psp, r3
+	isb
+
+	/* Check if we're returning to userspace */
+	and r1, lr, #4
+	ror r1, r1, #2
+	msr control, r1
+	isb
+	dsb
+
+	bx lr
+.size _interrupts_dispatch, .-_interrupts_dispatch
+.ltorg
+
+
+.globl _hal_invokePendSV
+.type _hal_invokePendSV, %function
+_hal_invokePendSV:
+	mov r1, #(1 << 28)
+	ldr r2, =ADDR_SCB
+	str r1, [r2, #4]
+	bx lr
+.size _hal_invokePendSV, .-_hal_invokePendSV
+.ltorg
+
+.globl hal_cpuReschedule
+.type hal_cpuReschedule, %function
+hal_cpuReschedule:
+	push {r0-r2, lr}
+	eor r0, r0, r0 /* default return value */
+	bl _hal_invokePendSV
+	pop {r1-r3, lr}
+	cmp r1, #NULL
+	beq hal_cpuReschedule0
+	push {r3-r4}
+	add r1, r1, #12
+
+spinlock:
+	ldrexb r3, [r1]
+	add r3, r3, #1
+	dmb
+	strexb r4, r3, [r1]
+	cmp r4, #0
+	bne spinlock
+	pop {r3-r4}
+hal_cpuReschedule0:
+	cpsie if
+	isb
+	dmb
+	bx lr
```

Comment:
What do you think about such a solution?
```suggestion
/* int hal_cpuReschedule(struct _spinlock_t *spinlock, spinlock_ctx_t *scp); */
/*  scp variable ( r1 register ) is not used, you do not need to keep it */
hal_cpuReschedule:
	push {r4, lr}
	mov r0, r4
	bl _hal_invokePendSV
	cmp r4, #NULL
	beq .Lhal_cpuReschedule0
	add r4, r4, #12

.Lspinlock:
	ldrexb r2, [r4]
	add r2, r2, #1
	dmb
	strexb r3, r2, [r4]
	cmp r3, #0
	bne .Lspinlock
.Lhal_cpuReschedule0:
	cpsie if
	isb
	dmb
	movs r0, #0 /* default return value */
	pop {r4, lr}
```
In addition, you can use `cbz` and `adds` instructions ( smaller resulting code ).

Reply:
I like your version, will update armv7m with it later, applied